File: web_ui/app/features/authentication/data/datasources/auth_remote_datasource.ts
import axios, { AxiosError } from "axios";
import { LoginUserDto } from "../dto/LoginUser.dto";
import { LoginDatasDto } from "../dto/LoginData.dto";
import axiosInstance from "@/app/core/interceptor";
import { config } from "@/app/core/config";
import { RegisterUserDto } from "../dto/RegisterUser.dto";

export class AuthRemoteDatasource {
  async loginUser(loginUserDto: LoginUserDto): Promise<LoginDatasDto> {
    let response;
    try {
      const rawData = {
        userName: loginUserDto.userName,
        password: loginUserDto.password,
      };

      response = await axiosInstance.post<{ data: any }>(
        `${config.quickChatService}/login`,
        rawData // axios will automatically stringify the body
      );

      if (response.status === 200 && response.data?.data) {
        return this.parseLoginData(response.data.data);
      }

      throw new Error("Invalid response format");
    } catch (error) {
      throw new Error(`${error}`);
    }
  }

  async registerUser(registerUserDto: RegisterUserDto): Promise<void> {
    let response;
    try {
      const rawData = {
        name: registerUserDto.name,
        userName: registerUserDto.userName,
        password: registerUserDto.password,
        password_confirmation: registerUserDto.password,
      };

      response = await axiosInstance.post<{ data: any }>(
        `${config.quickChatService}/register`,
        rawData
      );

      if (response.status === 201) {
        return;
      }

      throw new Error("Invalid response format");
    } catch (error) {
      throw new Error(`${error}`);
    }
  }

  // Private Methods
  //

  private parseLoginData(data: any): LoginDatasDto {
    console.log(`dita: ${data}`);
    // Add type validation if needed
    if (!data.token || !data.user) {
      throw new Error("Invalid login response format");
    }

    return {
      accessToken: data.accessToken,
      userId: data.userId,
      username: data.username,
      email: data.email,
      // Add any other fields from your LoginDataDto
      ...data,
    };
  }
}

File: mobile_ui/quickchat_ui/lib/features/authentication/data/datasources/auth_remote_datasource.dart
import 'dart:convert';
import 'dart:isolate';
import 'package:dartz/dartz.dart';
import 'package:dio/dio.dart';
import 'package:injectable/injectable.dart';
import 'package:quickchat_ui/config.dart';
import 'package:quickchat_ui/core/logic/server_exeption.dart';
import 'package:quickchat_ui/features/authentication/data/dto/LoginData.dto.dart';
import 'package:quickchat_ui/features/authentication/data/dto/LoginUser.dto.dart';
import 'package:quickchat_ui/features/authentication/data/dto/RegisterUser.dto.dart';

abstract class AuthRemoteDatasource {
  Future<Unit> connectServer();
  Future<LoginDatasDto> loginUser(LoginUserDto loginUserDto);
  Future<Unit> registerUser(RegisterUserDto registerUserDto);
}

@LazySingleton(as: AuthRemoteDatasource)
class AuthRemoteDatasourceImpl implements AuthRemoteDatasource {
  AuthRemoteDatasourceImpl({
    required this.dio,
    required this.config,
  });

  final Dio dio;
  final Config config;

  @override
  Future<Unit> connectServer() async {
    late Response<dynamic> response;

    try {
      response = await dio.head('${config.quickChatService}/test');
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 200) {
      return unit;
    } else {
      throw checkErrResponse(response);
    }
  }

  @override
  Future<LoginDatasDto> loginUser(LoginUserDto loginUserDto) async {
    late Response<dynamic> response;
    late LoginDatasDto resultData;

    try {
      final Map<String, dynamic> rawData = <String, dynamic>{
        'userName': loginUserDto.username,
        'password': loginUserDto.password,
      };

      // Send the login request
      response = await dio.post(
        '${config.quickChatService}/login',
        data: jsonEncode(rawData),
      );

      // Use Isolate.run for heavy computation (in this case, parsing the response)
      resultData = await Isolate.run(() {
        final Map<String, dynamic> result = response.data['data'];
        return LoginDatasDto.fromJson(result);
      });
    } on DioException catch (e) {
      throw checkErrResponse(e.response);
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 200) {
      return resultData;
    } else {
      throw checkErrResponse(response);
    }
  }

  @override
  Future<Unit> registerUser(RegisterUserDto registerUserDto) async {
    late Response<dynamic> response;

    try {
      final Map<String, dynamic> rawData = <String, dynamic>{
        'name': registerUserDto.name,
        'userName': registerUserDto.username,
        'password': registerUserDto.password,
        'password_confirmation': registerUserDto.password,
      };

      // Send the login request
      response = await dio.post(
        '${config.quickChatService}/register',
        data: jsonEncode(rawData),
      );
    } on DioException catch (e) {
      throw checkErrResponse(e.response);
    } catch (e) {
      throw checkErrResponse(response);
    }

    if (response.statusCode == 201) {
      return unit;
    } else {
      throw checkErrResponse(response);
    }
  }
}
